### PR TITLE
[Bugfix][MiniCPM-o] Bound native auto-response continuation

### PR DESCRIPTION
**Pull request by natureofnature**
## Purpose

MiniCPM-o native-duplex auto-response schedules another one-second silence unit whenever the model remains in `SPEAK`. If the model never samples `LISTEN`, that loop has no deterministic end: the final input is accepted, but `response.done` never arrives and the WebSocket remains active until the client timeout.

This change preserves long model-native replies while bounding each auto-response to 64 continuation units (about 64 seconds of silence-driven continuation):

- the first 63 continuation payloads retain their existing semantics;
- the 64th payload requests model-native termination with `force_listen`;
- a model `LISTEN` at the boundary is terminal and emits `response.done`;
- if the model ignores `force_listen`, the next continuation attempt emits a protocol `response.listen` plus exactly one `response.done`, without appending another unit;
- terminalization preserves the resumable stage-0 request, and an epoch/response-owner fence prevents a stale fallback from closing a response opened by a concurrent barge-in.

The existing 8-unit bound for non-auto-response continuation and other model paths are unchanged.

## Test Plan

**vLLM Version:** 0.27.0

**vLLM-Omni Commit:** `1f49ff7741eb87a62b98a975fc2bc9fca72d85bc`

Final regression suite in the remote vLLM 0.27.0 container:

```bash
PYTHONPATH=. pytest -o addopts= -q \
  tests/entrypoints/openai_api/test_duplex_handler.py
```

Full H800 lifecycle regression used the pending OmniInteract Nightly test with `#6587`, `#6626`, and this patch overlaid:

```bash
VLLM_OMNI_RUN_OMNIINTERACT_E2E=1 \
OMNIINTERACT_ROOT=/path/to/OmniInteract \
python3 -m pytest -o addopts= -s -v \
  tests/e2e/online_serving/test_minicpmo_4_5_duplex.py::test_minicpmo_4_5_omniinteract_nightly \
  --run-level full_model
```

## Test Result

- Final remote duplex-handler suite: `200 passed, 16 warnings in 15.09s`.
- Targeted Ruff check/format and repository safety hooks: passed.
- Before this patch, the shared-service 12-video run was `11/12`: the final input for `Breakfast_Burritos-7_135.mp4` was accepted, but the model remained in auto-response `SPEAK` until the WebSocket closed at 1063.05 seconds without `response.done` (`1 failed, 2 passed` pytest nodes in 3192.54 seconds).
- With the boundary enabled, the same H800 gate was `12/12` (`3 passed` pytest nodes in 2731.14 seconds). This earlier run exercised the model-honored `force_listen` path; the final head additionally covers the ignored-force hard stop and barge-in race with unit regressions.

| Subset | Cases | Mean TTFT | P99 TTFT | Mean TPOT |
| --- | ---: | ---: | ---: | ---: |
| 1q1a | 4/4 | 0.252 ms | 0.290 ms | 18.003 ms |
| 1q1a_math | 4/4 | 0.294 ms | 0.412 ms | 21.376 ms |
| 1qna | 4/4 | 0.222 ms | 0.265 ms | N/A |

This validates request completion and artifact production, not answer accuracy. The fixed playback horizon marked 1/4 1q1a, 1/4 1q1a_math, and 4/4 1qna WAVs as `audio_clipped`, so those cases were excluded from the official evaluation manifest rather than reported as accuracy-ready artifacts.

### CODE REVIEW

**Comment by natureofnature** [comment]
@amy-why-3459 PTAL

**Comment by amy-why-3459** [comment]
Request changes.

The diagnosis is right: native auto-response reschedules a 1s silence
unit while the model stays in SPEAK, and
`_native_response_continuations_remaining` returns True forever for
auto-response, so Breakfast_Burritos can accept the last input and
never emit `response.done`. The 8-unit cap for non-auto is untouched,
and the first 63 auto-response units keep the current payload. That
part is good.

The stated close path is not what the server does.

1. `force_listen` does not terminate an auto-response.

   Purpose says the 64th unit "terminates through the existing
   model-native LISTEN path". For auto-response that path is
   explicitly non-terminal:

   - `minicpmo45/data_plane.py` yields listen with
     `end_of_turn=False` and returns before any `turn_end` metadata
     is read.
   - `_send_one_native_duplex_event` treats
     `auto_responds + active response/request + end_of_turn is not
     True` as `non_terminal_auto_listen` and schedules another
     silence unit. It does not emit `response.done`.

   This PR only sets `payload["force_listen"] = True` when
   `count + 1 >= 64`. It does not stop scheduling, and it does not
   mark that listen as `end_of_turn`. After unit 64, `count + 1 >=
   64` stays true, so units 65+ keep going with `force_listen`.
   `_end_active_response_before_future_model_turn` also does not
   fire: continuation pins `duplex_turn_id` to the current
   `active_response_turn_id`.

   Please make the boundary actually close the request, e.g. both:
   - stop scheduling once `count >= 64` (same shape as the non-auto
     8-unit return);
   - treat a `force_listen`-triggered listen as terminal
     (`end_of_turn=True` → existing `response.listen` +
     `response.done`).

   A flag that only nudges logits is not a bound.

2. The new test only checks payload flags.

   `test_minicpmo_auto_response_continuation_uses_large_safety_boundary`
   asserts 64 appends, no flag on the first 63, and `force_listen`
   on the last. A 65th call still appends today. There is no
   `_send_one_native_duplex_event` case that a listen after the
   boundary emits `response.done`.

   Please add:
   - 65th `_maybe_continue_native_response` does not append (or,
     if you keep scheduling, a listen event must emit
     `response.done`);
   - an auto-response listen event that closes the response, not
     only a scheduler-payload test.

   Purpose's "final safety-boundary unit" is not true in the code:
   64 is when marking starts, not when continuation ends.

Should-fix:

- Keep a hard stop after the boundary. `force_listen` can be
  skipped by the once-per-`(request_id, seq)` logit guard, a
  failed silence schedule (`continuation_units` does not bump),
  or a chunk period that is not 1000 ms. Non-auto already stops
  scheduling at 8; auto-response needs that too.
- Explain 64. One unit is 1s, so this is ~64s of extra SPEAK
  before we even ask for LISTEN. How does that sit next to the
  non-auto 8-unit cap, #6587 playback horizon, and the Talker
  4096 rollover in #6626?
- `_native_response_continuations_remaining` still returns True
  for auto-response. Fine today (only the non-auto listen branch
  uses it), but the new cap is invisible to the helper.

The H800 12/12 is useful but it is #6587 + #6626 + this patch,
and the write-up already says the playback horizon marked several
WAVs `audio_clipped`. That does not prove this `force_listen` line
emits `response.done` on its own. Author self-review is still
missing.

**Comment by natureofnature** [comment]
Author self-review after rebasing onto current `main` and addressing the review:

- The boundary is now a hard stop, not only a `force_listen` hint: the 64th unit requests `LISTEN`, a model `LISTEN` closes the response, and an ignored hint causes the next continuation attempt to emit protocol `response.listen` plus exactly one `response.done` without appending again.
- The resumable request remains bound. The existing non-auto-response limit is unchanged.
- Fallback terminalization captures both epoch and response owner across its send await, so a barge-in that opens a new response cannot be closed by the stale fallback. A regression test exercises that exact interleaving.

Final remote vLLM 0.27.0 duplex-handler regression: `200 passed, 16 warnings in 15.09s`. Targeted Ruff/format and repository safety hooks pass.

**Comment by amy-why-3459** [comment]
Thanks — the previous blockers are fixed on 1f49ff7.

The 64th unit still asks for LISTEN. The 65th continuation no
longer appends: it emits protocol response.listen plus exactly
one response.done. A model LISTEN at the boundary is no longer
non_terminal_auto_listen once continuations_remaining is false,
so it hits the same close path and keeps the resumable request.
The helper and continue now share the auto 64 / non-auto 8
limits. The fallback snapshots epoch and response id across
the listen await; the barge-in interleaving test shows a new
response is not closed. Purpose matches the code. Self-review
is in.

Nits, not blocking:

1) After end_response(preserve_request=True), a late SPEAK on
   the same resumable request can begin_response again — the
   accepts-model-turn guard does not run when
   active_response_id is None. Breakfast_Burritos only needs
   the first response.done. An interactive client may see a
   second response.created. Please reject late SPEAK on that
   request/turn after the bound, and add a test.

2) The comment now places 64 against the 8-unit cap and the
   300s playback horizon. One line on how it sits next to the
   Talker 4096 rollover in #6626 would help.

I would approve.

**Review by amy-why-3459 — APPROVED** [review]
